sources/pam.c

```diff
@@ -0,0 +1,88 @@
+/*
+ * Odyssey.
+ *
+ * Scalable PostgreSQL connection pooler.
+*/
+
+#include <security/pam_appl.h>
+#include <kiwi.h>
+
+struct auth_request {
+    char *password;
+};
+
+
+static int
+pam_conversation(int msgc,
+		const struct pam_message **msgv,
+	       	struct pam_response **rspv,
+		void *authdata) {
+	struct auth_request *request = (struct auth_request *) authdata;
+	int counter, retcode;
+	
+	if (msgc < 1 || msgv == NULL || request == NULL) {
+		return PAM_CONV_ERR;
+	}
+	*rspv = malloc(msgc * sizeof(struct pam_response));
+	if (*rspv == NULL) {
+		return PAM_CONV_ERR;
+	}
+	
+	memset(*rspv, 0, msgc * sizeof(struct pam_response));
+	retcode = PAM_SUCCESS;
+	
+	for (counter = 0; retcode == PAM_SUCCESS && counter < msgc; counter++) {
+		if (msgv[counter]->msg_style == PAM_PROMPT_ECHO_OFF) {
+			(*rspv)[counter].resp = strdup(request->password);
+			if ((*rspv)[counter].resp == NULL) {
+				retcode = PAM_CONV_ERR;
+			}
+		}
+	}
+	
+	if (retcode != PAM_SUCCESS) {
+		for (counter = 0; counter < msgc; counter++)
+			free((*rspv)[counter].resp);
+		free(*rspv);
+	}
+	return retcode;
+}
+
+static int
+handle_pam_conversation(char *od_pam_service, char * user, char * password){
+	struct auth_request req = {
+		.password=password
+	};
+	struct pam_conv conv = {
+		pam_conversation,
+		.appdata_ptr = &req,
```

Comment:
Looks like you do not need this additional `struct auth_request` here and you can pass password pointer as is

Reply:
ok, fixed